### PR TITLE
Typescript target

### PR DESCRIPTION
**Pull request by ericvergnaud**
Candidate PR for the typescript target.

### CODE REVIEW

**Comment by ericvergnaud** [comment]
@parrt you can squash when merging

**Comment by Codex-** [comment]
Will be interesting to see if the types generated using this target are more accurate than the types inferred by `tsc` + `@types/antlr4`, looking forward to trying it out! 🚀

File: runtime/JavaScript/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
```
Comment (Codex-) [inline runtime/JavaScript/tsconfig.json:2]: I highly recommend adding `"strict": true,`:
https://www.typescriptlang.org/tsconfig#strict

**Comment by ericvergnaud** [comment]
@parrt looks good to me, all tests pass for all targets on latest dev

File: runtime/JavaScript/tsconfig.json (hunk 1 above)
Comment (ericvergnaud) [reply runtime/JavaScript/tsconfig.json:2]: We'll try that in the future

**Comment by ericvergnaud** [comment]
> Will be interesting to see if the types generated using this target are more accurate than the types inferred by `tsc` + `@types/antlr4`, looking forward to trying it out! 🚀

The types in this PR are not 'generated', they are hand coded

**Comment by Codex-** [comment]
> The types in this PR are not 'generated', they are hand coded

Apologies for the confusion, by generated I'm referring to the output generated by antlr when processing a grammar. 

Currently, we use the types package and then use typescript to infer and generate .d.ts files for the Parser, Listener, and Lexer that are generated for the Javascript runtime.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt [hunk 2]
```diff
@@ -10,8 +10,8 @@ Parser
 grammar T;
 s : e {<writeln("$e.v")>};
 e returns [int v]
-  : e '*' e     {$v = <Cast("BinaryContext","$ctx"):ContextMember({<Production("e")>(0)}, {<Result("v")>})> * <Cast("BinaryContext","$ctx"):ContextMember({<Production("e")>(1)}, {<Result("v")>})>;}  # binary
-  | e '+' e     {$v = <Cast("BinaryContext","$ctx"):ContextMember({<Production("e")>(0)}, {<Result("v")>})> + <Cast("BinaryContext","$ctx"):ContextMember({<Production("e")>(1)}, {<Result("v")>})>;}  # binary
+  : e '*' e     {$v = <Cast("BinaryContext","$ctx"):SubContextLocal({<Production("e")>(0)}, {<Result("v")>})> * <Cast("BinaryContext","$ctx"):SubContextLocal({<Production("e")>(1)}, {<Result("v")>})>;}  # binary
+  | e '+' e     {$v = <Cast("BinaryContext","$ctx"):SubContextLocal({<Production("e")>(0)}, {<Result("v")>})> + <Cast("BinaryContext","$ctx"):SubContextLocal({<Production("e")>(1)}, {<Result("v")>})>;}  # binary
```
Comment (KvanTTT) [inline runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt:14]: Sorry, but I don't understand why the changes related to test data are placed into the current pull request related to TypeScript. In my opinion they should be moved into another pull request.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg [hunk 3]
```diff
@@ -323,5 +323,8 @@ Declare_pred() ::= <<bool pred(bool v) {
 Invoke_pred(v) ::= <<this.pred(<v>)>>
 ParserTokenType(t) ::= "Parser.<t>"
 ContextRuleFunction(ctx, rule) ::= "<ctx>.<rule>"
+ContextListFunction(ctx, rule) ::= "<ctx>.<rule>()"
 StringType() ::= "String"
-ContextMember(ctx, subctx, member) ::= "<ctx>.<subctx>.<member>"
+ContextMember(ctx, member) ::= "<ctx>.<member>"
+SubContextLocal(ctx, subctx, local) ::= "<ctx>.<subctx>.<local>"
+SubContextMember(ctx, subctx, member) ::= "<ctx>.<subctx>.<member>"
```
Comment (KvanTTT) [inline runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg:330]: Changes related to test templates also should be extracted into another pull request.

**Review by KvanTTT — CHANGES_REQUESTED** [review]

**Comment by ericvergnaud** [comment]
They are part of this PR because they were required to make the typescript target tests runnable.
They don’t add any value in themselves, just a technical change to enable correct code generation of typescript tests.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt (hunk 2 above)
Comment (ericvergnaud) [reply runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt:14]: They are part of this PR because they were required to make the typescript target tests runnable.
They don’t add any value in themselves, it's simply a technical change to enable correct code generation of typescript tests.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg (hunk 3 above)
Comment (ericvergnaud) [reply runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg:330]: See previous comment

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt (hunk 2 above)
Comment (ericvergnaud) [reply runtime-testsuite/resources/org/antlr/v4/test/runtime/descriptors/LeftRecursion/MultipleAlternativesWithCommonLabel_1.txt:14]: In other words, they're not worth a PR unless we integrate the TypeScript target, so it makes sense to have them with this PR

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg (hunk 3 above)
Comment (parrt) [reply runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/CSharp.test.stg:330]: @ericvergnaud It would be nice to see the upgrade to the test specificity in a separate PR. Want me to pull those changes out into a PR we merge before your TypeScript target PR?

**Comment by parrt** [comment]
> @parrt looks good to me, all tests pass for all targets on latest dev

yay! I would like to see the test updates in a separate PR. happy to do that for you if you want. Then we can merge both.  

thanks for such a huge contribution!

**Comment by ericvergnaud** [comment]
yes please proceed to split the PR

you might split as follows actually:
 - test templates and descriptors
 - FileUtils.java
 - VisitorFile and ListenerFile (fixes incomplete PR #3878)
 - all the rest except maybe StructDecl, which I’m noticing now, there could be an incorrect merge with StructDecl, I’m 100% positive that I haven’t touched that stuff, so not sure why it shows up in the merge… also not sure how the go tests pass if that merge is incorrect

**Comment by KvanTTT** [comment]
> all the rest except maybe StructDecl, which I’m noticing now, there could be an incorrect merge with StructDecl, I’m 100% positive that I haven’t touched that stuff, so not sure why it shows up in the merge… also not sure how the go tests pass if that merge is incorrect

I highly recommend using [interactive rebase in IntelliJ IDEA](https://www.youtube.com/watch?v=bPX9VHjviEM) to filter out unrelated or incorrect changes and to prepare clear history.

**Comment by parrt** [comment]
Acknowledged @ericvergnaud @KvanTTT.  I'll see if intellij can help me tease this apart.

File: tool/pom.xml [hunk 4]
```diff
@@ -24,7 +24,7 @@
 		<dependency>
 			<groupId>org.antlr</groupId>
 			<artifactId>antlr4-runtime</artifactId>
-			<version>${project.version}</version>
+			<version>4.11.1</version>
```
Comment (parrt) [inline tool/pom.xml:27]: Wouldn't it be better to use the project.version not 4.11.1?

**Comment by parrt** [comment]
> all the rest except maybe StructDecl, which I’m noticing now, there could be an incorrect merge with StructDecl, I’m 100% positive that I haven’t touched that stuff, so not sure why it shows up in the merge… also not sure how the go tests pass if that merge is incorrect

yeah, weird. maybe it needed a rebase or something if you branched from dev earlier than these changes?   Hmm...i'll have to be careful pulling this apart.  I think looking at all changes and pulling those out manually is easier than merging interactively with ~180 commits haha.

**Comment by parrt** [comment]
@ericvergnaud yeah, it looks like you did a refactor/cleanup that whacked a previous changes in "make it easy to break" 974606ee447feb0b702e6ccc4994720b3513f62b   

Hmm...i wonder if i can do a rebase to make that merge properly....i wonder why a conflict didn't appear?

**Comment by ericvergnaud** [comment]
Agreed.
Seems we’re missing a PR for VisitiorFile/ListenerFile ? ((fixes incomplete PR #3878))
Then I’ll rebase and submit a new PR

File: tool/src/org/antlr/v4/codegen/model/ListenerFile.java [hunk 5]
```diff
@@ -61,7 +61,8 @@ public ListenerFile(OutputModelFactory factory, String fileName) {
 			}
 		}
 		ActionAST ast = g.namedActions.get("header");
-		if ( ast!=null ) header = new Action(factory, ast);
+		if ( ast!=null && ast.getScope() == null)
```
Comment (parrt) [inline tool/src/org/antlr/v4/codegen/model/ListenerFile.java:64]: just a reminder that I'd like { } around even single statements :)

Comment (ericvergnaud) [reply tool/src/org/antlr/v4/codegen/model/ListenerFile.java:64]: ;-) there wasn't one originally

Comment (parrt) [reply tool/src/org/antlr/v4/codegen/model/ListenerFile.java:64]: damn...has eroded. Maybe the rule is if it's one line like that but not two lines.

**Comment by ericvergnaud** [comment]
@parrt thanks for the intermediate PRs. I've cleaned up the suspicious changes, all changes left are valid. Xing fingers that the build succeeds. If it does you can squash and merge.

**Comment by parrt** [comment]
> If it does you can squash and merge.

acknowledged.

**Comment by ericvergnaud** [comment]
@parrt All green! :-)

**Comment by parrt** [comment]
I'll reset 4.11.2 -> 4.12.0 in milestones. thanks, @ericvergnaud great job!
